staker/execution_challenge_bakend.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2021-2022, Offchain Labs, Inc.
+// For license information, see https://github.com/nitro/blob/master/LICENSE
+
+package staker
+
+import (
+	"context"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/offchainlabs/nitro/validator"
+)
+
+type ExecutionChallengeBackend struct {
+	exec          validator.ExecutionRun
+	lastStep      validator.MachineStep
+	lastStepMutex sync.Mutex
+}
+
+// NewExecutionChallengeBackend creates a backend with the given arguments.
+// Note: machineCache may be nil, but if present, it must not have a restricted range.
+func NewExecutionChallengeBackend(executionRun validator.ExecutionRun) (*ExecutionChallengeBackend, error) {
+	return &ExecutionChallengeBackend{
+		exec: executionRun,
+	}, nil
+}
+
+func (b *ExecutionChallengeBackend) SetRange(ctx context.Context, start uint64, end uint64) error {
+	b.exec.PrepareRange(start, end)
+	return nil
+}
+
+func (b *ExecutionChallengeBackend) getStep(ctx context.Context, position uint64) (validator.MachineStep, error) {
+	b.lastStepMutex.Lock()
+	lastStep := b.lastStep
+	b.lastStepMutex.Unlock()
+	if lastStep != nil && lastStep.Ready() {
+		lastpos, err := lastStep.Position()
+		if err != nil && lastpos == position {
+
+			return lastStep, nil
```

Comment:
It made sense in this stage, because here machineStep had both hash and proof. Now that they're separate.. I agree lastStep only adds unnecessary complexity. I'll remove it in the last PR.